src/raftstore/store/peer.rs

```diff
@@ -598,9 +602,26 @@ impl Peer {
         Ok(())
     }
 
+    fn has_leader_lease(&self) -> bool {
+        // check whether current peer is raft leader in its lease
+        if !self.raft_group.raft.in_lease() {
+            return false;
+        }
+        // check the clock time against clock drift bound
+        let lower_clock_drift_bound =
+            (self.raft_group.raft.randomized_election_timeout as f64 *
+             (1.0 - self.election_timeout_clock_drift_bound_percentage)) as usize;
+        self.raft_group.raft.election_elapsed < lower_clock_drift_bound
+    }
+
     fn is_local_read(&self, req: &RaftCmdRequest) -> bool {
         if (req.has_header() && req.get_header().get_read_quorum()) ||
-           !self.raft_group.raft.in_lease() || req.get_requests().len() == 0 {
+           req.get_requests().len() == 0 {
+            return false;
+        }
+
+        // check whether local peer is raft leader and has valid lease for local read
+        if !self.has_leader_lease() {
```

Comment:
It's better to separate the checking for incoming requests and checking for the lease. Easier to write code/comments, and to read.